### ISSUE TITLE
这代码ok吗

### ISSUE THREAD

**Issue by 5464jmy** [AlanLi1997/slim-neck-by-gsconv#33]
x3 = torch.cat((x1, x2), dim=1)
b, n, h, w = x3 .size()
b_n = b * n // 2
y = x3 .reshape(b_n, 2, h * w)
y = y.permute(1, 0, 2)
y = y.reshape(2, -1, n // 2, h, w)
等于
y = torch.cat([x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]], dim=1)
这与两个卷积的对应通道数挨在一起不符合

**Comment by AlanLi1997** [comment]
@5464jmy 请注意，这行代码“y = torch.cat([x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]], dim=1)”并不等价于shuffle操作。torch.cat()输出的tensor往往并不严格按照输入的tensor排列(channel维度上)，这可能容易被忽视。或许你可以通过实验进一步验证以上描述。另外我不太理解“这与两个卷积的对应通道数挨在一起不符合”是要表达什么意思。
欢迎进一步讨论

**Comment by AlanLi1997** [comment]
@5464jmy 请注意，这行代码“y = torch.cat([x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]], dim=1)”并不等价于shuffle操作，“x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]”是四个tensor的整体的串接，并不是各tensor内channel的交替串接。
欢迎继续讨论。

**Comment by 5464jmy** [comment]
![45bb409747d3bad85ed170120b61a04](https://github.com/user-attachments/assets/09661885-bd6b-485a-b62c-a305c02c7e42)
各tensor内channel的交替串接是怎样的呢？
假如是（16,32,40,40）大小的x1, x2， 第一个batch x1,x2大小分别为（32,40,40），假定其中channel的排列方式如图中一样，
shuffle表达的是图中的序号一的操作吗，

**Comment by AlanLi1997** [comment]
@5464jmy “shuffle”操作的结果是图中②的样子；“y = torch.cat([x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]], dim=1)”的结果则是图中①的样子。随着channel的增加，①的shuffle强度会越来越弱，②的shuffle强度始终保持一致而不受channel数量变化的影响。

**Comment by 5464jmy** [comment]
class GSConv(nn.Module):
    # GSConv https://github.com/AlanLi1997/slim-neck-by-gsconv
    def __init__(self, c1, c2, k=1, s=1, p=None, g=1, d=1, act=True):
        super().__init__()
        self.c_ = c2 // 2
        self.cv1 = Conv(c1, self.c_, k, s, p, g, d, Conv.default_act)
        self.cv2 = Conv(self.c_, self.c_, 5, 1, p, self.c_, d, Conv.default_act)

    def forward(self, x):
        x1 = self.cv1(x)
        x2 = self.cv2(x1)
        x3 = torch.cat((x1, x2), dim=1)
        b, n, h, w = x3.size()
        b_n = b * n // 2
        y = x3.reshape(b_n, 2, h * w)
        y = y.permute(1, 0, 2)
        y = y.reshape(2, -1, n // 2, h, w)

        return (torch.cat((y[0], y[1]), 1), 
                torch.cat([x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]], dim=1), 
                torch.cat([x1.unsqueeze(2), x2.unsqueeze(2)], dim=2).flatten(1, 2))

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
a = GSConv(16, 64, k=3, s=2).to(device)
b = torch.rand((16, 16, 40, 40), device=device)
c1, c2, c3 = a(b)
print(c1.size())
print((c1 == c2).sum())
print((c1 == c3).sum())
-----------------------------------------------------------------------------------------------------------------------------------
torch.Size([16, 64, 20, 20])
tensor(409600, device='cuda:0')
tensor(25600, device='cuda:0')

我无法解释

**Comment by AlanLi1997** [comment]
@5464jmy 我注意到了，你使用的shuffle操作和[仓库中](https://github.com/AlanLi1997/slim-neck-by-gsconv/blob/a8f267b698942d9a7bda1f559ec9d64373daa67a/models_inPaper.py#L73-L76)不一致，根据你使用shuffle操作(可能是最早期的GSConv的shuffle版本)来说确实和“torch.cat([x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]], dim=1)”相同。
这是一段演示代码，运行后可以直观对比两种方法的差别，供你参考：

    import torch
    x1 = torch.randn(1,4,2,2)
    x2 = torch.randn(1,4,2,2)
    
    x3 = torch.cat((x1, x2), dim=1)
    # method 1 (last version)
    y1 = x3.reshape(x3.shape[0], 2, x3.shape[1] // 2, x3.shape[2], x3.shape[3])
    y1 = y1.permute(0, 2, 1, 3, 4)
    y1 = y1.reshape(y1.shape[0], -1, y1.shape[3], y1.shape[4])
    # method 2， equation yours shuffle version
    y2 = torch.cat((x1[:, ::2], x2[:, ::2], x1[:, 1::2], x2[:, 1::2]), dim=1)
    
    print(x1)
    print(x2)
    print(y1)
    print(y2)
    
    print(y1.size())
    print(y2.size())
    print(y1==y2)

**Comment by 5464jmy** [comment]
谢谢你的指正
